### PR TITLE
add opts to authorize so that pre_hooks can build more complex subjects/objects

### PR DESCRIPTION
**Pull request by hogiyogi597**
# **Description**

I'd like to encourage a separation of the `object` from other, potentially influential args that could be set dynamically. For example, in the context of a credit card application I may want to freeze a card and as part of that request I have the `card` to be frozen and the `frozen_reason` I'd like to set it to. It is reasonable to assume there are certain `frozen_reason`s that I would want to prevent users from using since they could be system level reasons.

Currently, `pre_hook`s only allow static values to be passed as options which is useful but I'd like to expand on that idea and allow callers of `authorize` to pass arguments that can be dynamically set through requests so it would be something like:

```elixir

def freeze_card(user, card, reason) do
  ...
  CardPolicy.authorize?(:card_freeze, user, card, reason)
  ...
end

# Policy
...
object :card do
  ...
  action :freeze do
    pre_hook :load_reason
    allow [:owns_card, :valid_reason]
  end
  ...
end

# Checks
def owns_card(%{id: id}, %{card: %{user_id: id}}), do: true
def owns_card(_, _), do: false

def valid_reason(_, %{reason: :frozen_by_user), do: true
def valid_reason(_, _), do: false

# Pre_hooks
def load_reason(user, card, reason), do: {user, %{card: card, reason: reason}}

```

# **Checklist**

- [x] I added tests that cover my proposed changes.
- [x] I updated the documentation related to my changes (if applicable).

### CODE REVIEW

File: lib/let_me/policy.ex [hunk 1]
```diff
@@ -1022,6 +1029,18 @@ defmodule LetMe.Policy do
...
+  You can achieve the same functionality dynamically using the opts on `authorize!`:
+
+      object :article do
+        action :view do
+          pre_hooks {MyApp.Policy.Checks, :set_age}
+          allow min_age: 50
+        end
+      end
+
+      MyApp.Policy.authorize!(:article_view, %{age: 10}, 50)
```
Comment (woylie) [inline lib/let_me/policy.ex:1041]: The object argument is missing here, and the last argument needs to be an enumerable.

File: test/support/hooks.ex [hunk 2]
```diff
@@ -1,11 +1,17 @@
 defmodule LetMe.TestHooks do
   @moduledoc false
 
-  def preload_handsomeness(%{} = subject, %{} = object, factor: factor) do
-    {Map.put(subject, :handsomeness, subject.id * factor), object}
+  def preload_handsomeness(%{} = subject, %{} = object, opts) do
+    factor = Keyword.get(opts, :factor)
```
Comment (woylie) [inline test/support/hooks.ex:5]:
```suggestion
    factor = Keyword.fetch!(opts, :factor)
```

File: lib/let_me/policy.ex [hunk 3]
```diff
@@ -376,8 +376,11 @@ defmodule LetMe.Policy do
 
   The error reason can be customized by setting the `:error_reason` option when
   using the module.
+
+  The last parameter is a set of arguments that can be defined dynamically
+  which will be passed into any `pre_hook`s defined on the resource's policy.
   """
-  @callback authorize(atom, any, any) :: :ok | {:error, any}
+  @callback authorize(atom, any, any, any) :: :ok | {:error, any}
```
Comment (woylie) [inline lib/let_me/policy.ex:383]: With your changes of the `build_pre_hook_calls` macro, the fourth argument is expected to be an `Enumerable.t()`. But I also think we can save us the `Enum.into` and enforce a `keyword`.

**Review by woylie — CHANGES_REQUESTED** [review]
Thanks, this looks useful, just a few comments. Also, there are some warnings now about invalid references in the docs, can you fix them?

Do you think it would make sense to pass those options to the check functions as well?

**Comment by hogiyogi597** [comment]
Thanks for your comments! I'll fix up the branch with your suggestions. 

Also, I appreciate you working with me through contributions. I hope you don't feel like I am infringing on your project! If you do, just let me (lol) know. I want to be respectful of your work. I have really enjoyed working on our LetMe implementation.

> Do you think it would make sense to pass those options to the check functions as well?

I originally thought that it might make sense to pass into the check functions but I could go either way honestly. I kinda like how simple the check functions interface is currently. That's why I initially opted to just allow pre hooks to build up the subject/object. But I am open to ideas; what would be your preference?

File: CHANGELOG.md [hunk 4]
```diff
@@ -43,7 +43,7 @@
 
 ### Changed
 
-- Renamed `c:LetMe.Policy.authorized?/3` to `c:LetMe.Policy.authorize?/3`,
+- Renamed `c:LetMe.Policy.authorized?/3` to `c:LetMe.Policy.authorize?/4`,
```
Comment (woylie) [inline CHANGELOG.md:46]:
```suggestion
- Renamed `c:LetMe.Policy.authorized?/3` to `c:LetMe.Policy.authorize?/3`,
```

File: lib/let_me/policy.ex [hunk 5]
```diff
@@ -376,15 +376,18 @@ defmodule LetMe.Policy do
 
   The error reason can be customized by setting the `:error_reason` option when
   using the module.
+
+  The last parameter is a set of arguments that can be defined dynamically
+  which will be passed into any `pre_hook`s defined on the resource's policy.
   """
-  @callback authorize(atom, any, any) :: :ok | {:error, any}
+  @callback authorize(atom, any, any, Keyword.t()) :: :ok | {:error, any}
```
Comment (woylie) [inline lib/let_me/policy.ex:383]:
```suggestion
  @callback authorize(atom, any, any, keyword) :: :ok | {:error, any}
```

File: lib/let_me/policy.ex [hunk 6]
```diff
@@ -395,14 +398,14 @@ defmodule LetMe.Policy do
       iex> MyApp.Policy.authorize!(:article_update, user_2, article)
       ** (LetMe.UnauthorizedError) unauthorized
   """
-  @callback authorize!(atom, any, any) :: :ok
+  @callback authorize!(atom, any, any, any) :: :ok
```
Comment (woylie) [inline lib/let_me/policy.ex:401]:
```suggestion
  @callback authorize!(atom, any, any, keyword) :: :ok
```

File: lib/let_me/policy.ex [hunk 7]
```diff
@@ -413,7 +416,7 @@ defmodule LetMe.Policy do
       iex> MyApp.Policy.authorize?(:article_update, user_2, article)
       false
   """
-  @callback authorize?(atom, any, any) :: boolean
+  @callback authorize?(atom, any, any, any) :: boolean
```
Comment (woylie) [inline lib/let_me/policy.ex:419]:
```suggestion
  @callback authorize?(atom, any, any, keyword) :: boolean
```

**Review by woylie — APPROVED** [review]

**Comment by woylie** [comment]
That should do, thanks!

**Pull request merged**